### ISSUE TITLE
Menu: Menu button with Icon are not well centered

### ISSUE THREAD

**Issue by enguerran** [cozy/cozy-ui#338]
For https://github.com/cozy/cozy-contacts, I need a menu with dots icon as the main button, so I wrote:

```jsx
  <Menu
    position="right"
    component={
      <Button theme="secondary" extension="narrow">
        <Icon icon="dots" />
      </Button>
    }
  >
    <MenuItem icon={<Icon icon="delete" />} onSelect={deleteAction.action}>
      {deleteAction.label}
    </MenuItem>
  </Menu>
```

as you can see in the below screenshot, it is not well centered:

![](https://i.imgur.com/vBTsbPo.png)

**Comment by enguerran** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [(2) ETU je peux supprimer un contact depuis sa fiche](https://trello.com/c/eD8KjGbM/724-2-etu-je-peux-supprimer-un-contact-depuis-sa-fiche)

**Comment by GoOz** [comment]
Yeah, it's a `margin-right` on the Icon. It was thought for button with both Icon and text.
I'll figure somthing out. Thanks.

**Comment by ptbrowne** [comment]
Without the fix, cozy-drive and cozy-photos cannot use the `<Menu />`, it seems a priority to me to fix this.

**Comment by GoOz** [comment]
So it turns out it's not that easy.

So far `<Button />` just takes `children` as its content, so basically UI has no control over what's inside. 
We managed the margin between an icon and the text by using a CSS selector on the `svg` element because the icon is always one.
But, if this way worked in the case of Icon + text, it doesn't work when there's just an icon on it's own, the `svg` keeps its margin, and there's no way to tell when it's alone.

The only clean solution I can find, that doesn't require any writing of CSS from the developer or a mandatory DOM, is to remove the children and only work from props like `label` and `icon`.

For example, today we use `<Button />` like this:
```jsx
<Button theme='secondary'>
    <Icon icon="delete" /> Text
</Button>
```
And we would have to use it like this then:
```jsx
<Button theme='secondary' icon='delete' label='Text' />
```

This is a breaking change, so we'll add a depreciation warning.

Does anyone have a different use case for Buttons that wouldn't fit or require some other changes?

**Comment by goldoraf** [comment]
@GoOz It looks like it's the best solution, and the solution adopted by a good number of libraries, for instance: http://react-toolbox.io/#/components/button. So :+1:, even if it means to have to make a lot of changes (again) on the app side.

**Comment by ptbrowne** [comment]
To easy the transition path, jscodeshift could be used ;)

http://astexplorer.net/#/gist/cc94c9c4cc446d1958729300581bb250/384a77b1f9657d081b65128f0a778d5ee0f66651

**Comment by y-lohse** [comment]
I'll make a few comments, but I haven't actually looked at the offending code so feel free to tell me off.

First, I think it was misguided to have a rule altering the presentation of a child element that you don't control. Something like

```
button svg
    margin-right .5rem
```

is not great. But hey, that's easy to say in hindsight. Let's get rid of it.


But just sayin', I think we could fix it with the [`:only-child` selector](https://developer.mozilla.org/fr/docs/Web/CSS/:only-child).

```
button svg
    margin-right .5rem

button svg:only-child
    margin-right 0
```

Terrible idea, let's not do this.

Finally, I can see trouble with only a `label` prop. Just an example off the top of my head where we'll want a more complicated label with **2** icons (probably):

<img width="205" alt="capture d ecran 2018-02-15 17 58 44" src="https://user-images.githubusercontent.com/2261445/36269646-f0037da0-1279-11e8-99c9-c4d33f0a1704.png">

So I think `<Button theme='secondary' icon='delete' label='Text' />` is ok, as long `<Button theme='secondary'>Whoop Whoop <DropDown /></Button>` also works.

Peace out ☮️

**Comment by ptbrowne** [comment]
The only-child idea is interesting, why do you say it is terrible ?


I dont think the dropdown opener should be a button. It is too strongly coupled with the selector behavior. For example it has an opening state, the icon inside should maybe rotate to indicate that when opened, clicking again will close the menu.

**Comment by GoOz** [comment]
@y-lohse We can't use `:only-child` or `:first-child:last-child` because text node doesn't count as a node.
In this example…
```jsx
<Button>
   <svg />
   text
</Button>
```
…as far as the DOM is concerned, SVG **is** the only child it has, it juste ignores text nodes. There is absolutely no way of knowing what's inside unless you control it which is the purpose of the `icon` and `label` props. When we have this, we can be sure what would be the children and then use `:only-child`.

As for you dropdown use case, I agree with @ptbrowne, it think it should be a component on its own. `<Button>` is only meant for standard `button`, this would be way out of its league. I don't think complicating things because they look a like is the best thing to do.
And nothing forbids to use the same style on different components.

**Comment by y-lohse** [comment]
> We can't use :only-child or :first-child:last-child because text node doesn't count as a node.

Aw you're right, sucks.

For the record, I didn't mind the `:only-child`, but I mind arbitrarily changing every `svg` inside a `Button`. So I wasn't keen on building things on top of that.

______

You might be right about that particular example of the dropdown, but my general point was that I'm pretty sure we'll want more than just a string as a label. It could be a word in bold, or a number of other things.
I'm fine with the `icon` and `label` prop, let's just not loose the ability to pass childrens to the `Button`!

**Comment by GoOz** [comment]
Well I guess we could let `children` to open some doors, allow the developer to "hack" the first purpose of the component, but in this case, it's the developer responsability to style what inside.
I'm cool with it.

**Comment by ptbrowne** [comment]
OK, it would be cool to see thanks to jscodeshift the number of case how many cases of "hacked" buttons we have in our current codebases to have an approximate idea of if it is useful or not. I am OK to let `children` for now but I think it would be useful to run the jscodeshift to convert the current buttons to label/icon. Thanks for the feedback @y-lohse !

**Comment by GoOz** [comment]
Fixed by #357
